basic backend

flupe opened the pull request:
A very crude backend translating Agda's internal syntax to core syntax. Constructors get eta-expanded to have them fully-applied in core.

Only functions allowed should have a single clause with no patterns, meaning avoid implicit arguments for now (eagerly inserted by Agda as function parameters).

~~Will rebase myself on current main~~ (done!)

Usage:

```
$ cabal run agda-core -- -i test test/one.agda -vagda-core:30
```

jespercockx commented:
> Will rebase myself on current main (done!)

I'm afraid another rebase/merge is needed.

jespercockx reviewed (approved):
Nice work!

flupe commented on app/Main.hs:
```diff
@@ -1,4 +1,112 @@
+                    <+> text "to core syntax:" <+> text e
+      Right ty -> reportSDoc "agda-core.check" 5 $ text "Type:" <+> text (show ty)
+
+    case theDef def of
```
Since I don't know how we're gonna interact with the type checker yet, I'd rather keep translation of terms and checking of definitions separate for now.

flupe commented:
Replaced `String` with `Doc`, rebased on `main`, I think it's ready to go.
Also had to add `Agda +debug` to a `cabal.project` file for debug logs to show up when compiling with cabal with latest Agda.
